Add validation for Kubernetes object name fields and ProtocolType

youngnick opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:

1. If this is your first time contributing to Gateway API, please read our
   developer guide (https://gateway-api.sigs.k8s.io/devguide/)
   and our community page (https://gateway-api.sigs.k8s.io/community/).
2. If this is your first time contributing to a Kubernetes project, please read
   our contributor guidelines:
   https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution
3. Please label this pull request according to what type of issue you are
   addressing, especially if this is a release targeted pull request. For
   reference on required PR/issue labels, read here:
   https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
4. If you want *faster* PR reviews, read how:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design
/kind gep

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup

**What this PR does / why we need it**:

This PR ensures that there are no required fields that allow empty values, and that all required fields have consistent validation.

It does this by introducing a new string type alias, ObjectName, for Kubernetes object names, and by ensuring that the Listener ProtocolType has validation.

On the plus side, that was all the required types that could be zero I could find. The Port number field already has validation to ensure it's between 1 and 65535.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #867 

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->
```release-note
Validation for Kubernetes object names has been updated
Listener ProtocolType now has validation
```

howardjohn reviewed (commented):
Sorry if my regex knowledge is wrong here but I think this breaks the examples for ExternalService types where we have things like google.com?

youngnick commented:
I have assumed that the ObjectReferences here must be to Kubernetes objects. I can see that it would be handy to be able to fudge a direct reference to a non-Kubernetes thing, but I guess I had assumed that in that case there would be some Kubernetes object to represent the external object first.

If we are not going to require that references are to Kubernetes objects, then we should record that fact alongside the all the object name references.

hbagdi commented on apis/v1alpha2/gateway_types.go:
```diff
@@ -247,6 +244,10 @@ type Listener struct {
 // protocols. Any protocol defined by implementations will fall under custom
 // conformance.
+//
+// +kubebuilder:validation:MinLength:=1
+// +kubebuilder:validation:MinLength:=255
```
```suggestion
// +kubebuilder:validation:MinLength=1
// +kubebuilder:validation:MinLength=255
```

hbagdi commented:
+1 to both points by Nick.

youngnick commented:
Dang, I should have talked about this in the meeting today. I'll start a Slack thread so we can go async.

jpeach commented:
> I have assumed that the ObjectReferences here must be to Kubernetes objects.

Yeh, I had the same assumption.

youngnick commented:
Okay, I'll leave this until we can come to some agreement about the project's stance on if the references must be to Kubernetes objects only, https://kubernetes.slack.com/archives/CR0H13KGA/p1632187311091400 is the Slack thread to discuss.

robscott commented:
The Slack thread referenced by @youngnick above has gotten fairly long. Here's a summary of what I think the current state is for each of the two relevant questions.

**1. Should we validate Name in ObjectRefs?**
This is nuanced, and I don't think there's consensus here yet. My personal opinion is that we should not. Each Kubernetes resource can define its own name validation (example: [EndpointSlice](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/discovery/validation/validation.go#L54-L58), [others](https://cs.k8s.io/?q=ValidateObjectMeta%5C(&i=nope&files=validation.go&excludeFiles=&repos=kubernetes/kubernetes)). Although the validation proposed here currently represents the least restrictive validation I'm aware of among current k8s objects, there is no guarantee that will be true in the future. The corresponding [docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names) highlight the validation described here as the most common name validations, but do not provide any guarantees that resources will always fall into these buckets. The lack of upstream validation for ObjectRef names makes me think we should avoid this.

**2. Should we be open to using ObjectRefs to reference non-k8s objects?**
There are significant concerns raised in the Slack thread about using ObjectRefs to refer to non-k8s objects. @hbagdi summarizes it well with the following:

>This API is complex enough making extension points more loosely defined and arbitrary could have effects that are hard for us to foresee today.

Although I'm personally open to the idea of referencing virtual or non-k8s resources, I don't feel as strongly about it. It seems mostly reasonable to start without this concept with the idea that we can add it later if necessary with a simple addition to docs and no change in validation.

youngnick commented:
Okay, finishing up for the day but tomorrow after the meeting I will update this and remove the object name validation, and clarify that objects must be Kubernetes resources.

youngnick commented:
In the latest commit, I:
- removed the validation on the ObjectName field
- Updated docs on all the reference fields with "must reference a valid Kubernetes object" language
- Added examples for ProtocolType
- Updated the regex for ProtocolType to match the given examples

youngnick commented:
Well, the system works, turned out there were a couple of regex problems, which the tests picked up! Fixed now.

robscott commented on apis/v1alpha2/shared_types.go:
```diff
@@ -298,6 +302,14 @@ type Group string
+//
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=253
+type ObjectName string
```
I [followed up with sig-api-machinery](https://groups.google.com/g/kubernetes-sig-api-machinery/c/V-qnyODuMt4) and it sounds like there is a [minimum set of validation we can provide here](https://github.com/kubernetes/kubernetes/blob/b290671b1b28ab4af8164e2b69b334fab584558e/staging/src/k8s.io/apimachinery/pkg/api/validation/path/name.go#L30-L46). It may be a bit difficult to translate into regex validation though, and others recommended against adding validation here. Although that minimum set of validation does not include a max length, 253 seems like a reasonable value to me based on the most common validation + ensure we don't end up with an unbounded list.

robscott reviewed (commented):
Thanks for the work on this @youngnick! I personally don't think we _need_ regex validation for ObjectName, but also would be fine if it was added. I'm leaving a hold on here in case you want to add something based on the [sig-api-machinery thread](https://groups.google.com/g/kubernetes-sig-api-machinery/c/V-qnyODuMt4).

/lgtm
/hold

youngnick commented:
I don't think that doing that level of checking is worthwhile when the apiserver will stop you creating objects with that name anyway; I think it's more important to get the "must refer to an existing type" clauses in.

/hold cancel

robscott commented:
Thanks!

/lgtm
